diff --git a/src/minicom.c b/src/minicom.c
index 2719f8cce5a3edf42b34918a870299004d813d21..06dd7be5840dc2fca733a6d0c995e52f814ca568 100644
--- a/src/minicom.c
+++ b/src/minicom.c
@@ -173,7 +173,6 @@ static void shjump(int sig)
 static ELM *mc_getline(WIN *w, int no)
 {
   int i;
-  static ELM outofrange[MAXCOLS] = {{0,0,0}};
 
   if (no < us->histlines) {
     /* Get a line from the history buffer. */
@@ -188,13 +187,20 @@ static ELM *mc_getline(WIN *w, int no)
   /* Get a line from the "us" window. */
   no -= us->histlines;
   if (no >= w->ys) {
-    if (outofrange[0].value == 0) {
-      for (i = 0; i < MAXCOLS; i++) {
-        outofrange[i].value = ' ';
-        outofrange[i].color = us->color;
-        outofrange[i].attr  = us->attr;
+    static int alloced_columns;
+    static ELM *outofrange;
+    int cols = w->x2 + 1;
+    if (cols > alloced_columns) {
+      free(outofrange);
+      outofrange = malloc(sizeof(*outofrange) * cols);
+      assert(outofrange);
+      alloced_columns = cols;
+
+      for (i = 0; i < cols; i++) {
+	outofrange[i].value = i == 0 ? '~' : ' ';
+	outofrange[i].color = us->color;
+	outofrange[i].attr  = us->attr;
       }
-      outofrange[0].value = '~';
     }
     return outofrange;
   }
